plugin/plugins/trace/service/service.go

```diff
@@ -0,0 +1,280 @@
+// in the above the naming schema
+func (tss *TraceStreamingService) ListenEndBlock(
+	ctx sdk.Context,
+	req abci.RequestEndBlock,
+	res abci.ResponseEndBlock,
+) error {
+	eventId := int64(1)
+	eventTypeId := 1
+
+	// write req
+	key := fmt.Sprintf(LogMsgFmt, tss.currentBlockNumber, EndBlockEvent, eventId, RequestEventType, eventTypeId)
+	if err := tss.writeEventReqRes(ctx, key, &req); err != nil {
+		return err
+	}
+
+	// write state changes
+	if err := tss.writeStateChange(ctx, EndBlockEvent, eventId); err != nil {
+		return err
+	}
+
+	// write res
+	key = fmt.Sprintf(LogMsgFmt, tss.currentBlockNumber, EndBlockEvent, eventId, ResponseEventType, eventTypeId)
+	if err := tss.writeEventReqRes(ctx, key, &res); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
+// in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. Setting this to `false` will give fire-and-forget semantics.
+// When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
+// be replayed to all listeners when the node restarts and all successful listeners that received data
+// prior to the halt will receive duplicate data.
+func (tss *TraceStreamingService) HaltAppOnDeliveryError() bool {
+	return tss.haltAppOnDeliveryError
+}
+
+// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
+// Do we need this and an intermediate writer? We could just write directly to the buffer on calls to Write
+// But then we don't support a Stream interface, which could be needed for other types of streamers
+func (tss *TraceStreamingService) Stream(wg *sync.WaitGroup) error {
+	if tss.quitChan != nil {
+		return errors.New("`Stream` has already been called. The stream needs to be closed before it can be started again")
+	}
+	tss.quitChan = make(chan struct{})
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-tss.quitChan:
+				return
+			case by := <-tss.srcChan:
+				tss.stateCacheLock.Lock()
+				tss.stateCache = append(tss.stateCache, by)
+				tss.stateCacheLock.Unlock()
+			}
+		}
+	}()
+	return nil
+}
+
+// Close satisfies the io.Closer interface
+func (tss *TraceStreamingService) Close() error {
+	close(tss.quitChan)
+	return nil
+}
+
+func (tss *TraceStreamingService) writeStateChange(ctx sdk.Context, event string, eventId int64) error {
+	// write all state changes cached for this stage
+	tss.stateCacheLock.Lock()
+	kodec := tss.codec.(*codec.ProtoCodec)
+	kvPair := new(types.StoreKVPair)
+	for i, stateChange := range tss.stateCache {
+		key := fmt.Sprintf(LogMsgFmt, tss.currentBlockNumber, event, eventId, StateChangeEventType, i+1)
+		if err := kodec.UnmarshalLengthPrefixed(stateChange, kvPair); err != nil {
+			return err
+		}
+		if err := tss.writeEventReqRes(ctx, key, kvPair); err != nil {
+			return err
+		}
```

Comment:
If either of these error conditions are triggered, tss.stateCacheLock is never unlocked, resulting in deadlock.

Reply:
fixed.